xugrid/ugrid/burn.py

```diff
@@ -0,0 +1,300 @@
+    vertices = np.vstack([exterior] + interiors).astype(np.float64)
+    triangles = mapbox_earcut.triangulate_float64(vertices, rings).reshape((-1, 3))
+    triangle_indices, grid_indices = grid.celltree._locate_faces(vertices, triangles)
+    if all_touched:
+        return grid_indices
+    else:
+        centroids = grid.centroids[grid_indices]
+        inside = points_in_triangles(
+            points=centroids,
+            face_indices=triangle_indices,
+            faces=triangles,
+            vertices=vertices,
+        )
+        return grid_indices[inside]
+
+
+def _burn_polygons(
+    polygons: "geopandas.GeoSeries",  # type: ignore # noqa
+    like: "xugrid.Ugrid2d",
+    values: np.ndarray,
+    all_touched: bool,
+    output: FloatArray,
+) -> None:
+    exterior_coordinates = [
+        shapely.get_coordinates(exterior) for exterior in polygons.exterior
+    ]
+    interior_coordinates = [
+        [shapely.get_coordinates(p_interior) for p_interior in p_interiors]
+        for p_interiors in polygons.interiors
+    ]
+    to_burn = np.empty(like.n_face, dtype=bool)
+
+    for exterior, interiors, value in zip(
+        exterior_coordinates, interior_coordinates, values
+    ):
+        to_burn = _locate_polygon(like, exterior, interiors, all_touched)
+        output[to_burn] = value
+
+    return
+
+
+def _burn_points(
+    points: "geopandas.GeoSeries",  # type: ignore # noqa
+    like: "xugrid.Ugrid2d",
+    values: np.ndarray,
+    output: FloatArray,
+) -> None:
+    """
+    Simply searches the points in the ``like`` 2D topology.
+    """
+    xy = shapely.get_coordinates(points)
+    to_burn = like.locate_points(xy)
+    output[to_burn] = values
+    return
+
+
+def _burn_lines(
+    lines: "geopandas.GeoSeries",  # type: ignore # noqa
+    like: "xugrid.Ugrid2d",
+    values: np.ndarray,
+    output: FloatArray,
+) -> None:
+    """
+    This algorithm breaks any linestring down into edges (two x, y points). We
+    search and intersect every edge in the ``like`` grid, the intersections are
+    discarded.
+    """
+    xy, index = shapely.get_coordinates(lines, return_index=True)
+    # From the coordinates and the index, create the (n_edge, 2, 2) shape array
+    # containing the edge_coordinates.
+    linear_index = np.arange(index.size)
+    segments = np.column_stack([linear_index[:-1], linear_index[1:]])
+    # Only connections with vertices with the same index are valid.
+    valid = np.diff(index) == 0
+    segments = segments[valid]
+    edges = xy[segments]
+    # Now query the grid for these edges.
+    edge_index, face_index, _ = like.intersect_edges(edges)
+    # Find the associated values.
+    line_index = index[1:][valid]
+    value_index = line_index[edge_index]
+    output[face_index] = values[value_index]
+    return
+
+
+def burn_vector_geometry(
+    gdf: "geopandas.GeoDataframe",  # type: ignore # noqa
+    like: "xugrid.Ugrid2d",
```

Comment:
I was a bit in doubt about the name ``like`` here, conceptually I perceived this method different than ``regrid_like`` or ``full_like``. As it is returning a copy of ``like`` however, it is essentially doing such a thing, but it feels a bit unintuitive. I cannot think of a better name though. ``grid`` doesn't cover it (as UgridDataset is also allowed), and ``template`` introduces suddenly new nomenclature to the package.  Just sharing my thoughts, you can resolve without notice.